packages/glob/src/internal-pattern.ts

```diff
@@ -0,0 +1,227 @@
+import * as assert from 'assert'
+import * as path from 'path'
+import * as pathHelper from './internal-path-helper'
+import {Minimatch, IMinimatch, IOptions as IMinimatchOptions} from 'minimatch'
+import {MatchResult} from './internal-match-result'
+import {Path} from './internal-path'
+
+const IS_WINDOWS = process.platform === 'win32'
+
+export class Pattern {
+  comment: boolean = false
+  negate: boolean = false
+  searchPath: string
+  private minimatch: IMinimatch
+  private rootRegExp: RegExp
+  private trailingSlash: boolean = false
+
+  constructor(pattern: string) {
+    pattern = pattern || ''
+
+    // Comment
+    if (pattern.startsWith('#')) {
+      this.comment = true
+      this.searchPath = (undefined as unknown) as string
```

Comment:
Why not allow `searchPath` and the below to be nullable types?

Reply:
i'll change to empty string. Not sure if i'm doing something wrong, but I just find the not-null constraints of typescript annoying rather than helpful. If it's nullable then the compiler requires i use the nullable deference everywhere (`?.`) which i then have to coalesce even when i know it's never null. Which just makes the code less readable, and production behavior less clear.

Reply:
now that comment isnt a property this problem went away (as you suggested on the other comment)

Reply:
good call btw